lib/lz4.c

```diff
@@ -1178,8 +1178,25 @@ LZ4_FORCE_INLINE int LZ4_decompress_generic(
         const BYTE* match;
         size_t offset;
 
-        /* get literal length */
         unsigned const token = *ip++;
+
+        /* shortcut for common case */
+        /* this shortcut was tested on x86 and x64, where it improves decoding speed.
+         * it has not yet been benchmarked on ARM, Power, mips, etc. */
+        if (((ip + 14 + 2 <= iend) & (op + 14 + 16 <= oend))
+          & ((token < 15*16) & ((token & 0xF) <= 12)) ) {
```

Comment:
Can you use `(token < (15 << 4))`, so it is more clear that we are looking at the top 4 bits.

Reply:
Was it faster to only copy 16 byte matches, instead of the possible 19?

Reply:
16 bytes is meant to match an SSE register,
or two 64-bits registers.

The maximum possible match length in the token is 18 bytes, since 19 requires reading additional bytes.
I haven't tested speed for 18 bytes.
I expect it to perform worse as it requires an additional register copy all the time, while a match length of 17 or 18 is likely marginal.

Reply:
I just tested a variant using `matchLength <= 18` on x64.
Performance was similar on my small sample set (`silesia`, `calgary`, `enwik8`), with differences within margin error.

Reply:
Tested on godbolt : 
- clang trunk uses SSE registers : 
- gcc 7.2 doesn't, its assembly output looks more complex

https://godbolt.org/g/5cn7t4

Reply:
`memcpy(,,18)` introduces a mix of SSE and normal registers : 
https://godbolt.org/g/fiWFgM